src/txn/txn_ckpt.c

```diff
@@ -181,18 +181,30 @@ __checkpoint_apply(WT_SESSION_IMPL *session, const char *cfg[],
 	int (*op)(WT_SESSION_IMPL *, const char *[]))
 {
 	WT_DECL_RET;
-	u_int i;
+	u_int i, j;
 
 	/* If we have already locked the handles, apply the operation. */
 	for (i = 0; i < session->ckpt_handle_next; ++i) {
 		if (session->ckpt_handle[i] == NULL)
 			continue;
 		WT_WITH_DHANDLE(session, session->ckpt_handle[i],
 		    ret = (*op)(session, cfg));
-		WT_RET(ret);
+		WT_ERR(ret);
 	}
 
-	return (0);
+err:
+	/*
+	 * If we have an error somewhere in processing the handles, then
+	 * we need to mark earlier trees dirty.
+	 */
+	if (ret != 0)
+		for (j = 0; j < i; ++j) {
```

Comment:
I'm not convinced this is always adequate. I believe it's necessary to mark any tree that any phase of checkpoint has visited as dirty, not just handles that have been visited in this phase of checkpoint.

By phase, I mean any place that calls ``__checkpoint_apply``.